packages/firestore/lite/src/api/reference.ts

```diff
@@ -49,3 +55,222 @@ export class DocumentReference<T = firestore.DocumentData>
+  }
+
+  endAt(
+    docOrField: unknown | firestore.DocumentSnapshot<unknown>,
+    ...fields: unknown[]
+  ): firestore.Query<T> {
+    // TODO(firestorelite): Implement
+    throw new Error('Not implemented');
+  }
+
+  endBefore(
+    docOrField: unknown | firestore.DocumentSnapshot<unknown>,
+    ...fields: unknown[]
+  ): firestore.Query<T> {
+    // TODO(firestorelite): Implement
+    throw new Error('Not implemented');
+  }
+
+  withConverter<U>(
+    converter: firestore.FirestoreDataConverter<U>
+  ): firestore.Query<U> {
+    return new Query<U>(this.firestore, this._query, converter);
+  }
+}
+
+export class CollectionReference<T = firestore.DocumentData> extends Query<T>
+  implements firestore.CollectionReference<T> {
+  constructor(
+    readonly firestore: Firestore,
+    readonly _path: ResourcePath,
+    readonly _converter?: firestore.FirestoreDataConverter<T>
+  ) {
+    super(firestore, InternalQuery.atPath(_path), _converter);
+  }
+
+  get id(): string {
+    return this._query.path.lastSegment();
+  }
+
+  get path(): string {
+    return this._query.path.canonicalString();
+  }
+
+  withConverter<U>(
+    converter: firestore.FirestoreDataConverter<U>
+  ): firestore.CollectionReference<U> {
+    return new CollectionReference<U>(this.firestore, this._path, converter);
+  }
+}
+
+export function collection(
+  firestore: FirebaseFirestore,
+  collectionPath: string
+): CollectionReference<firestore.DocumentData>;
+export function collection(
+  reference: DocumentReference,
+  collectionPath: string
+): CollectionReference<firestore.DocumentData>;
+export function collection(
+  parent: firestore.FirebaseFirestore | firestore.DocumentReference<unknown>,
+  relativePath: string
+): CollectionReference<firestore.DocumentData> {
+  if (relativePath.length === 0) {
+    throw new FirestoreError(
+      Code.INVALID_ARGUMENT,
+      `Invalid path (${relativePath}). Empty paths are not supported.`
+    );
+  }
+
+  const path = ResourcePath.fromString(relativePath);
+  if (parent instanceof Firestore) {
+    if (DocumentKey.isDocumentKey(path)) {
+      throw new FirestoreError(
+        Code.INVALID_ARGUMENT,
+        `Invalid path (${path}). Path points to a document.`
+      );
+    }
+    return new CollectionReference(parent, path);
+  } else {
+    const doc = hardCast(parent, DocumentReference);
+    const absolutePath = doc._key.path.child(path);
+    if (DocumentKey.isDocumentKey(absolutePath)) {
+      throw new FirestoreError(
+        Code.INVALID_ARGUMENT,
+        `Invalid path (${absolutePath}). Path points to a document.`
+      );
+    }
+    return new CollectionReference(doc.firestore, absolutePath);
+  }
+}
+
+export function doc(
+  firestore: FirebaseFirestore,
+  documentPath: string
+): DocumentReference<firestore.DocumentData>;
+export function doc<T>(
+  reference: CollectionReference<T>,
+  documentPath?: string
+): DocumentReference<T>;
+export function doc<T>(
```

Comment:
Combines https://osscs.corp.google.com/firebase-sdk/firebase-js-sdk/+/master:packages/firestore/src/api/database.ts;l=2274 and https://osscs.corp.google.com/firebase-sdk/firebase-js-sdk/+/master:packages/firestore/src/api/database.ts;l=568